ext/openssl/ossl_ts.c

```diff
@@ -0,0 +1,1527 @@
+ *
+ * Mandatory parameters that need to be set in the Factory:
+ * * Factory#serial_number
+ * * Factory#gen_time
+ *
+ * In addition one of either Request#policy_id or Factory#default_policy_id
+ * must be set.
+ *
+ * Raises a TimestampError if creation fails.
+ *
+ * call-seq:
+ *       factory.create_timestamp(key, certificate, request) -> Response
+ */
+static VALUE
+ossl_tsfac_create_ts(VALUE self, VALUE key, VALUE certificate, VALUE request)
+{
+    VALUE serial_number, def_policy_id, gen_time, additional_certs;
+    VALUE str, cert;
+    STACK_OF(X509) *inter_certs;
+    VALUE ret;
+    EVP_PKEY *sign_key;
+    X509 *tsa_cert;
+    TS_REQ *req;
+    TS_RESP *response = NULL;
+    TS_RESP_CTX *ctx = NULL;
+    BIO *req_bio;
+    const char * err_msg = NULL;
+    int i;
+
+    tsa_cert = GetX509CertPtr(certificate);
+    sign_key = GetPrivPKeyPtr(key);
+    req = GetTsReqPtr(request);
+
+    if (!(ctx = TS_RESP_CTX_new())) {
+	err_msg = "Memory allocation failed.";
+	goto end;
+    }
+    serial_number = ossl_tsfac_get_serial_number(self);
+    if (serial_number == Qnil) {
+	err_msg = "@serial_number must be set.";
+	goto end;
+    }
+    gen_time = ossl_tsfac_get_gen_time(self);
+    if (gen_time == Qnil) {
+	err_msg = "@gen_time must be set.";
+	goto end;
+    }
+    def_policy_id = ossl_tsfac_get_default_policy_id(self);
+    if (def_policy_id == Qnil && !TS_REQ_get_policy_id(req)) {
+	err_msg = "No policy id in the request and no default policy set";
+	goto end;
+    }
+
+    TS_RESP_CTX_set_serial_cb(ctx, ossl_tsfac_serial_cb, &serial_number);
+    if (!TS_RESP_CTX_set_signer_cert(ctx, tsa_cert)) {
+	err_msg = "Certificate does not contain the timestamping extension";
+	goto end;
+    }
+
+    additional_certs = ossl_tsfac_get_additional_certs(self);
+    if (additional_certs != Qnil) {
+	if (!(inter_certs = sk_X509_new_null())) {
+	    err_msg = "Memory allocation failed.";
+	    goto end;
+	}
+	if (rb_obj_is_kind_of(additional_certs, rb_cArray)) {
+	    for (i = 0; i < RARRAY_LEN(additional_certs); i++) {
+		cert = rb_ary_entry(additional_certs, i);
+		sk_X509_push(inter_certs, GetX509CertPtr(cert));
+	    }
+	}
+	else {
+	    sk_X509_push(inter_certs, GetX509CertPtr(additional_certs));
+	}
+	// this dups the sk_X509 and ups each cert's ref count
+	TS_RESP_CTX_set_certs(ctx, inter_certs);
+	sk_X509_free(inter_certs);
+    }
+
+    TS_RESP_CTX_set_signer_key(ctx, sign_key);
+    if (def_policy_id != Qnil && !TS_REQ_get_policy_id(req)) {
+	ASN1_OBJECT *def_policy_id_obj = obj_to_asn1obj(def_policy_id);
+	TS_RESP_CTX_set_def_policy(ctx, def_policy_id_obj);
+	ASN1_OBJECT_free(def_policy_id_obj);
+    }
+    if (TS_REQ_get_policy_id(req))
+	TS_RESP_CTX_set_def_policy(ctx, TS_REQ_get_policy_id(req));
+    TS_RESP_CTX_set_time_cb(ctx, ossl_tsfac_time_cb, &gen_time);
+
+    TS_RESP_CTX_add_md(ctx, EVP_get_digestbyname(OBJ_nid2sn(NID_md5)));
+    TS_RESP_CTX_add_md(ctx, EVP_get_digestbyname(OBJ_nid2sn(NID_sha1)));
```

Comment:
Would it be okay to take an opinionated stance and not support SHA1 and MD5?

Reply:
I started wondering where the list comes in the first place. I suspect the best thing would be to not have a hard-coded list and to let the user set their own one.

Reply:
:+1: I'll refactor this to allow the user to specific a list.

Reply:
I added a `Factory#allowed_digests` accessor in 5e45f1306e99c498541af11f2b5747ed7dc9a561. `Factory#create_timestamp` now requires the list of allowed digest algorithms to be provided.